Split Value::Float into 32 and 64 bit variants

tomhoule opened the pull request:
Casting between f32 and f64 can cause precision loss. I have a minimal reproduction here [in this repo](https://github.com/tomhoule/mysql-async-float-precision-loss-repro). This happens for us when writing and reading from `float4` columns. I haven't been able to find the proper place to test this, so there are no tests, but I manually confirmed that this fixes the issue in our case.

The naming of the variants is just an idea, I chose what rust users are probably most familiar with.

This is a breaking change, but I think it's pretty important to fix.

blackbeam commented:
Hi. Let's discuss this issue a bit more.

As far as I can see casting between f32 and f64 did cause precision loss [here][1], where `6.4` (which is `f64`) explicitly casted to `f32` (the result is 6.400000095367432). Precision loss should go away if this cast is removed.

[1]: https://github.com/tomhoule/mysql-async-float-precision-loss-repro/blob/master/src/main.rs#L9

tomhoule commented:
Yes indeed you're right, in this case. I think the precision loss still happens if the cast is done inside the database, which is the issue we encountered. I'll adjust my example tomorrow and see, but it shouldn't affect the result.

tomhoule commented:
Ok I'm back on my work computer now :) I pushed a commit to the repro repo that better shows the issue - the cast happens inside the database, and without the type information on our side we have no way to know whether we should cast it back to f32 or not. It's not a problem if you have the knowledge of the column type statically, but for tools that have to work across arbitrary databases schemas, it's an issue.

blackbeam commented:
First of all let me state, that i want to accept this PR (but with `F32` renamed to `Float` and `F64` renamed to `Double` to match with mysql internals documentation). The reason is that now i see, that [this implicit conversion][1] is dumb. The behavior should be:

1.  For writes:
    1.  when user writes `f32` to `FLOAT` column - `f32` must be serialized to `Value::Float` and sent to the server.
    2.  when user writes `f64` to `FLOAT` column - `f64` must be serialized to `Value::Double` and sent to the server. Server side cast is user's fault and there is nothing we can do with this. If schema is unknown to user, then he must avoid this loss dynamically (blocked by blackbeam/mysql_async#106)
    3.  when user writes `f32` to `DOUBLE` column - `f32` must be serialized to `Value::Float` and sent to the server. Server side cast will be lossless.
    4.  when user writes `f64` to `DOUBLE` column - `f64` must be serialized to `Value::Double` and sent to the server.
2. For reads:
    1.  when user reads `f32` from `FLOAT` column - column value must be deserialized as `Value::Float`. No cast will be performed.
    2.  when user reads `f64` from `FLOAT` column - column value must be deserialized as `Value::Float`. Lossless cast to `f64` will be performed.
    3.  when user reads `f32` form `DOUBLE` column - column value must be deserialized as `Value::Double` but conversion must be denied because it'll lead to precision loss ([allowed][1] at the moment). If schema is unknown to user, then he must avoid this path dynamically.
    4.  when user reads `f64` from `DOUBLE` column - column value must be deserialized as `Value::Double`. No cast will be performed.

I see now that your case is 1.ii + unknown schema. This means, that you are blocked by blackbeam/mysql_async#106, so I'll fix it in the nearest future.

BTW: Do you have time/desire to continue work on this PR (i.e. rename variants and alter the behavior as stated above)?

[1]: https://github.com/blackbeam/rust_mysql_common/blob/master/src/value/convert/mod.rs#L361-L391

tomhoule commented:
Yes I can do this - I think I see where to do this in the codebase. I'm not too familiar with the testing setup of the crate though, is there a place where meaningful tests could be written for this?

blackbeam commented:
Great!. Crate uses `cargo test`. Tests can go into [this module](https://github.com/blackbeam/rust_mysql_common/blob/master/src/value/convert/mod.rs#L1173).

tomhoule commented:
Ok I've adjusted and tested for the read path, and renamed the variants. If that looks good to you, I'll just wait for https://github.com/blackbeam/mysql_async/issues/106 to do the write path. I'm not sure what the API will look like though, is the expected column type going to be passed in [`ComStmtExecuteRequestBuilder::add_param()`](https://github.com/blackbeam/rust_mysql_common/blob/caf66461e8a25383cbe9c9a4e26de87bd7e22077/src/packets.rs#L1303)?

blackbeam commented:
> I'll just wait for blackbeam/mysql_async#106 to do the write path.

Great. blackbeam/mysql_async#106 is fixed, but, unfortunately, won't fix the issue stated in 1.ii, because it turns out that parameters field of a statement packet [contains dummy data][1] 😔. This means, that if schema is unknown, then some kind of schema analysis will be required to avoid the issue.

> I'm not sure what the API will look like though, is the expected column type going to be passed in ComStmtExecuteRequestBuilder::add_param()?

At the moment of statement execution each `Value` must be of the right type (user's responsibility). So the required change is trivial, viz. `match` expressions ([this one][2] and [this one][3]) must be altered to match proper behavior for new `Float` and `Double` variants, i.e:

*   MySql types for `Float` and `Double` must be `MYSQL_TYPE_FLOAT` and `MYSQL_TYPE_DOUBLE` correspondingly,
*   `Float` and `Double` should be written using `write_f32::<LE>` and `write_f64::<LE>` correspondingly.

Thanks again for your work!

[1]: https://github.com/mysql/mysql-server/blob/ea7d2e2d16ac03afdd9cb72a972a95981107bf51/sql/sql_prepare.cc#L44
[2]: https://github.com/blackbeam/rust_mysql_common/blob/caf66461e8a25383cbe9c9a4e26de87bd7e22077/src/packets.rs#L1309
[3]: https://github.com/blackbeam/rust_mysql_common/blob/caf66461e8a25383cbe9c9a4e26de87bd7e22077/src/io.rs#L60

tomhoule commented:
Ok I pushed again, I think this time it's ready for review :) Thanks for the guidance!

blackbeam commented:
I'm pushing this forward to fix blackbeam/mysql_async#107
Thanks!
